### PR TITLE
fix: files.add with pull streams + ability to seek

### LINKED ISSUE #1317: files.add with pull stream does not work
- **Version**: 0.28.0
- **Platform**: macOS Node.js 9.11.1
- **Subsystem**: files

#### Description:

Using `files.add(pull stream)` does not work

#### Steps to reproduce the error:

1. add with pull stream (promise)

    ```js
    const pull = require('pull-stream')
    const Ipfs = require('ipfs')
    const ipfs = new Ipfs

    ipfs.on('ready', () => {
        ipfs.files.add(pull.values([Buffer.from('test')]))
          .then(console.log)
          .catch(console.error)
    })
    ```

    output:

    ```
      /Users/alan/Desktop/test/ipfs-pull.test.js:31
          .then(console.log)
          ^

      TypeError: Cannot read property 'then' of undefined
          at IPFS.ipfs.on (/Users/alan/Desktop/test/ipfs-pull.test.js:31:5)
          at IPFS.emit (events.js:180:13)
          at done (/Users/alan/Desktop/test/node_modules/ipfs/src/core/boot.js:61:10)
          at /Users/alan/Desktop/test/node_modules/async/internal/parallel.js:39:9
          at /Users/alan/Desktop/test/node_modules/async/internal/once.js:12:16
          at replenish (/Users/alan/Desktop/test/node_modules/async/internal/eachOfLimit.js:59:25)
          at iterateeCallback (/Users/alan/Desktop/test/node_modules/async/internal/eachOfLimit.js:49:17)
          at /Users/alan/Desktop/test/node_modules/async/internal/onlyOnce.js:12:16
          at /Users/alan/Desktop/test/node_modules/async/internal/parallel.js:36:13
          at done (/Users/alan/Desktop/test/node_modules/ipfs/src/core/components/start.js:20:7)
    ```

2. add with pull stream (callback)

    ```js
    const pull = require('pull-stream')
    const Ipfs = require('ipfs')
    const ipfs = new Ipfs

    ipfs.on('ready', () => {
        ipfs.files.add(pull.values([Buffer.from('test')]), console.log)
    })
    ```

    output:

    ```
    Error: Invalid arguments, data must be an object, Buffer or readable stream
        at Function.add.promisify (/Users/alan/Desktop/test/node_modules/ipfs/src/core/components/files.js:199:25)
        at Object.add (/Users/alan/Desktop/test/node_modules/promisify-es6/index.js:32:27)
        at IPFS.ipfs.on (/Users/alan/Desktop/test/ipfs-pull.test.js:46:14)
        at IPFS.emit (events.js:180:13)
        at done (/Users/alan/Desktop/test/node_modules/ipfs/src/core/boot.js:61:10)
        at /Users/alan/Desktop/test/node_modules/async/internal/parallel.js:39:9
        at /Users/alan/Desktop/test/node_modules/async/internal/once.js:12:16
        at replenish (/Users/alan/Desktop/test/node_modules/async/internal/eachOfLimit.js:59:25)
        at iterateeCallback (/Users/alan/Desktop/test/node_modules/async/internal/eachOfLimit.js:49:17)
        at /Users/alan/Desktop/test/node_modules/async/internal/onlyOnce.js:12:16
    ```

---

Note that the following cases are **successful**:

1. addPullStream with chunks of { content: [pull stream] }
2. add with [{ content: [pull stream] }]

Files spec (which allows these combinations) for reference: https://github.com/ipfs/interface-ipfs-core/blob/master/SPEC/FILES.md

### PR DESCRIPTION
**Pull request by alanshaw**
fixes #1317

tested by https://github.com/ipfs/interface-ipfs-core/pull/256

### CODE REVIEW

**Comment by diasdavid** [comment]
I understand that this PR needs the latest of js-ipfs-api. I'm having a hard time releasing js-ipfs-api. Too many timeouts. Wanna take a look at that?

**Comment by alanshaw** [comment]
Sure, I'll take a look now

**Comment by alanshaw** [comment]
Ok so, for `js-ipfs-api`:

1. The CircleCI build is failing because it's using Node.js 10 and the native modules are failing to compile
2. Locally, if using Node 9.9.0+ all the tests fail because of [this issue in `ipfsd-ctl`](https://github.com/ipfs/js-ipfsd-ctl/issues/223)
3. `name.publish` just takes ages, I've submitted a PR to increase the timeout on this test here https://github.com/ipfs/js-ipfs-api/pull/749

**Comment by diasdavid** [comment]
@alanshaw something is not working properly in this PR.

**Comment by alanshaw** [comment]
According to the CI the failing tests are to do with `files.cat`...not something that this PR should have effected 😟 - I'll take a look.

File: package.json [hunk 1]
```diff
@@ -73,7 +73,7 @@
     "expose-loader": "^0.7.5",
     "form-data": "^2.3.2",
     "hat": "0.0.3",
-    "interface-ipfs-core": "^0.61.0",
+    "interface-ipfs-core": "^0.64.2",
```
Comment (diasdavid) [inline package.json]: @alanshaw something might have changed here

**Comment by achingbrain** [comment]
@diasdavid CI seems happy with this now but some of our (unrelated to this PR) tests time out occasionally.

**Pull request merged**